### ISSUE TITLE
Type inconsistency between Array sizes and Table number of rows

### ISSUE THREAD

**Issue by cquike** [casacore/casacore#1037]
The number of rows in a Table is of type rownr_t. However the size of an Array has the type size_t. Since the later can be used to retrieve the values from a given column, at the very least the maximum size of the array should be larger than the maximum number of table rows.

As an example, here is the interface provided at ScalarColumn.h:

```
    // Get the vector of all values in the column.
    // According to the assignment rules of class Array, the destination
    // vector must be empty or its length must be the number of cells
    // in the column (i.e. the number of rows in the table).
    void getColumn (Vector<T>& vec, Bool resize = False) const;

    // Get the vector of all values in the column.
    Vector<T> getColumn() const;
```
In a sense, the move to rownr_t has alleviated the problem, since now, for 64 bits platforms, the two types are equivalent. However in a general case the two types can potentially be different. A possible solution is to typedef rownr_t size_t. That means that for 32bits  platforms the tables cannot handle 64 bits indexes. But it makes CASA internally consistent.

**Comment by cquike** [comment]
Sorry,
just correction myself. The fact that Arrays can also be used in the putColumn method:
```
    // Put the vector of all values in the column.
    // The length of the vector must be the number of cells in the column
    // (i.e. the number of rows in the table).
    void putColumn (const Vector<T>& vec);

```
means that the maximum size of array length and table row numbers have to be the same. The possible inescapable outcome is that rownnr_t must be defined as size_t.

**Comment by aroffringa** [comment]
That doesn't make sense to me. Making rownr_t a size_t would make 64-bit row numbers completely inaccessible on 32-bit platforms, including get/put by rownr which currently work, as well as getColumn/putColumn. Currently, that functionality is available, except that it is not possible to completely get/fill a column at once from an Array. The latter seems not like an incredible interesting case when the table has >2^31 rows, since it would be several 10s of GB, which wouldn't fit on a 32-bit machine anyway.

Casacore should however report an error when getColumn is used on a >2^31 set, if it doesn't do that yet.

So, making rownr_t a size_t would break get/put for >2^31 rows altogether.
Making rownr_t always int64_t (current) makes only the getColumn/putColumn not work for >2^31 rows. But they would still not work with rownr_t=size_t. So, making rownr_t a size_t makes it only worse.

I've noticed Python arrays can not properly handle > 2^31 (32?) elements in an array (at least last time I tried), so also the Python interface breaks to some level with >2^31 rows.

**Comment by cquike** [comment]
I guess that it is another point of seen it: place the proper checks. There are probably quite a number of interfaces  to fix with some interesting challenges.

For all the classes that derive from BaseColumn:

```
    virtual void BaseColumn::getScalarColumn (ArrayBase& dataPtr) const;
    virtual void BaseColumn::getArrayColumn (ArrayBase& dataPtr) const;
    virtual void BaseColumn::getColumnSlice (const Slicer&, ArrayBase& dataPtr) const;
    virtual void BaseColumn::getScalarColumnCells (const RefRows& rownrs,
				                                                  ArrayBase& dataPtr) const;
    virtual void BaseColumn::putScalarColumn (const ArrayBase& dataPtr);
    virtual void BaseColumn::putColumnSlice (const Slicer&, const ArrayBase& dataPtr);
    virtual void BaseColumn::putScalarColumnCells (const RefRows& rownrs,
				                                                  const ArrayBase& dataPtr);

```
 Thei idea of not being able to use those methods to fill/retrieve the data at once is interesting. But it has its own challenges: should it throw an exception in that case? Also it imposes a non-signifcant burden for applications that use casacore and try to be portable: in all cases in which data needs to be read/written at once the application must have the provisions to cope for potential failure. That's a non trivial change.


 Then there are a number of interfaces that take a vector of number rows. In this case, basic table manipulatations cannot be done if the types of Vector and table sizes don't match:

```
    void BaseTable::removeRow (const Vector<rownr_t>& rownrs);
    void BaseTable::removeRow (const Vector<uInt>& rownrs);

    BaseTable* BaseTable::select (const Block<Bool>& mask);
    BaseTable* BaseTable::project (const Block<String>& columnNames);

    virtual BaseTable::Vector<rownr_t> rowNumbers() const;
    virtual BaseTable::Vector<rownr_t>* rowStorage();
    virtual Bool BaseTable::adjustRownrs (rownr_t nrrow, Vector<rownr_t>& rownrs,
			       Bool determineOrder) const;
    void ColumnsIndex::fillRowNumbers (Vector<rownr_t>& rows, rownr_t start, rownr_t end) const;
    void ColumnsIndexArray::fillRowNumbers (Vector<rownr_t>& rows, rownr_t start, rownr_t end,
	              	       Bool unique) const;
    void ColumnsIndexArray::fillRownrs (rownr_t npts, const Block<rownr_t>& nrel);
    RefRows::RefRows (const Vector<rownr_t>& rowNumbers, Bool isSliced = False,
             Bool collapse = False);
    const Vector<rownr_t>& RefRows::rowVector() const
    virtual Vector<rownr_t> RefTable::rowNumbers() const;  
    Vector<rownr_t>  RefTable::rootRownr (const Vector<rownr_t>& rownrs) const;
    virtual Vector<rownr_t>* RefTable::rowStorage();
    virtual Bool RefTable::adjustRownrs (rownr_t nrrow, Vector<rownr_t>& rownrs,
			                                       Bool determineOrder) const;
    RowNumbers::RowNumbers (const Vector<rownr_t>& rows)
    TableProxy TableProxy::selectRows (const Vector<Int64>& rownrs,
	                                   		 const String& outName);

```

 And additionally there are some internal variables that are also affected, so I don't think it is currently possible to practically work with 64bits tables in a 32bits environment:

```
    ConcatRows:   Block<rownr_t>  itsRows;
    RefRows:        Vector<rownr_t> itsRows;
   RefRowsSliceIter:    Vector<rownr_t> itsRows;
   RefTable:          Vector<rownr_t> rowStorage_p

```

**Comment by aroffringa** [comment]
Indeed, applications that want to be portable and support "64-bit MSes" should not use these functions that write entire columns. They cannot use them to begin with, because they can't store the data in 32-bit addressed memory. The getColumn-methods should indeed probably throw an exception in that case, if they don't already do that. A quick safety guard might be to not allow 64-bit tables on 32-bit machines altogether, since as you point out a few other methods would also probably not work as expected. If 64-bit tables with 32-bit machine support is important, the checks can be made for specific methods, such that iterating over a table would still be possible, which is probably the most important use-cases for calibration/imaging tasks.

**Comment by cquike** [comment]
Hi,
I just want to mention that this is in the context of CASA, which is not the kind of one-purpose only application running on a single hardware. So I am looking at this from the perspective of long term maintenance of a "relatively" generic application.
In that sense, I see more as a programming interface issue. Basically, by having two different types, casacore is signalling that these two types _are different_. So imaging the following CASA code (slightly modified to remove noise): 

```
     Table myTable;
      rownr_t nrows = myTable.nrow();
      Cube<Float> wtsp(nCorrelations(),nChannels(),nrows,0.0);

      for (size_t irow=0; irow < nrows ;++irow) {
         for (size_t icorr=0; icorr < nCorrelations(); ++icorr) {
             wtsp(icorr),0,irow) = someValue;
         }
     }

```       
Here an Array is being filled with the values of a table. Incidentally, this code is inside the iterator, so myTable is actually a RefTable coming from one iteration of a TableIterator. In CASA each iteration is controlled by the grouping parameters, so there also no easy way to control whether the table that you get in one iteration is within the 32 bits integer space.
 Since rownr_t and size_t are different types the code above has a potential overflow. Similarly the reversed operation (filling a column from an array) can potentially overflow (RISC-V architecture, for instance, can be implemented with 128 bits word-width, which means size_t is 128 bits there).
If one is to write that code in a fully compliant way it would be rather cumbersome. And there is a lot of code like that in CASA...
